### PR TITLE
go sqlgen: Add support for converting "Zero" value fields to NULL in the db

### PR DESCRIPTION
**Pull request by willhug**
Summary: Adds support to sqlgen to convert "Zero" value model fields to NULL in 
the database.  This only needs to be applied when inserting a value to the 
database because when we read from the db NULL values are noop (leaving the 
zero value).

I put some cleanups in the PR as well, would be easy to pull them out if we want.

Additionally, I wrote tests for the "Main" use cases, but I'm very open 
to add tests for other edge cases, let me know and I'll add them.

### CODE REVIEW

File: internal/fields/sql.go [hunk 1]
```diff
@@ -77,6 +77,10 @@ func (f Valuer) Value() (driver.Value, error) {
 			return iface.MarshalJSON()
 		}
 		return json.Marshal(i)
+	case f.Tags.Contains("zeroIsNull"):
```
[the remark this answers is not in the document]
Comment (berfarah) [reply internal/fields/sql.go]: As per offline convo: `implicitnull`?

File: internal/fields/sql.go [hunk 2]
```diff
@@ -115,8 +119,7 @@ var _ driver.Valuer = &Valuer{}
 // into the type dictated by our descriptor.
 type Scanner struct {
 	*Descriptor
-	value   reflect.Value
-	isValid bool
```
[the remark this answers is not in the document]
Comment (willhug) [reply internal/fields/sql.go]: Yup!

File: sqlgen/zeronull_test.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
+package sqlgen
+
+import (
```
[the remark this answers is not in the document]
Comment (willhug) [reply sqlgen/zeronull_test.go]: Yeah, adding some tests here and seeing edge cases, will update PR

**Comment by willhug** [comment]
Adding some tests for filters to make sure they still work properly.

File: internal/fields/sql.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (berfarah) [reply internal/fields/sql.go]: I would prefer all lowercase since that's how go's stdlib does it (`omitempty`).

**Comment by willhug** [comment]
Updated!  Also added a db test thing I've been using locally to get tests to run without an EnvVar every time.

File: internal/testfixtures/db.go [hunk 4]
```diff
@@ -28,6 +28,8 @@ var DefaultDBConfig = DBConfig{
 func init() {
 	if pw := os.Getenv("DB_PASSWORD"); pw != "" {
 		DefaultDBConfig.Password = pw
+	} else {
```
Comment (berfarah) [inline internal/testfixtures/db.go]: We can change the password to just always be dev and update the docker container on here?

File: sqlgen/integration_test.go [hunk 5]
```diff
@@ -201,10 +204,11 @@ func Benchmark(b *testing.B) {
...
 	user := &User{
-		Id:   1,
-		Name: "Bob",
-		Uuid: testfixtures.CustomType{'1', '1', '2', '3', '8', '4', '9', '1', '2', '9', '3'},
-		Mood: &mood,
+		Id:           1,
+		Name:         "Bob",
+		Uuid:         testfixtures.CustomType{'1', '1', '2', '3', '8', '4', '9', '1', '2', '9', '3'},
+		Mood:         &mood,
+		ImplicitNull: "test",
```
Comment (berfarah) [inline sqlgen/integration_test.go]: Hmm - do we want this in our benchmark?

File: sqlgen/implicitnull_test.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	newRow.Id = 1
+
+	var rows []*ImplicitNull
+	require.NoError(t, db.Query(context.Background(), &rows, nil, nil))
+	assert.Equal(t, []*ImplicitNull{
+		newRow,
+	}, rows)
+}
+
+func GetCount(db *sql.DB, countQuery string) (int, error) {
```
Comment (stephen) [inline sqlgen/implicitnull_test.go]: nit: don't export this?

File: internal/fields/sql.go [hunk 7]
```diff
@@ -77,6 +77,10 @@ func (f Valuer) Value() (driver.Value, error) {
 			return iface.MarshalJSON()
 		}
 		return json.Marshal(i)
+	case f.Tags.Contains("implicitnull"):
+		if IsZero(f.value) {
```
Comment (stephen) [inline internal/fields/sql.go]: i'm excited by how small this change is that i missed it in my first read.

File: internal/fields/reflect.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"reflect"
+)
+
+// IsZero returns whether the value of the passed in reflect value is the
+// zero value of that type.
+// Mostly copied from https://github.com/golang/go/issues/7501
+// Works for mostly everything except zero initialized arrays (e.g.
+// `var ray [5]string` will not be "Zero")
+func IsZero(v reflect.Value) bool {
```
Comment (stephen) [inline internal/fields/reflect.go]: nit: should we make this private?

File: sqlgen/implicitnull_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	timeField := time.Date(2000, 10, 3, 0, 0, 0, 0, time.UTC)
+	newRow := &ImplicitNull{
+		NullStr:   "hello",
+		NullInt:   123,
+		NullFloat: 1.23,
+		NullBool:  true,
+		NullByte:  []byte("hello there"),
+		NullTime:  timeField,
+	}
+	if _, err := db.InsertRow(context.Background(), newRow); err != nil {
```
Comment (stephen) [inline sqlgen/implicitnull_test.go]: could we add a test for `UpdateRow` as well?

File: internal/fields/sql.go [hunk 10]
```diff
@@ -77,6 +77,10 @@ func (f Valuer) Value() (driver.Value, error) {
 			return iface.MarshalJSON()
 		}
 		return json.Marshal(i)
+	case f.Tags.Contains("implicitnull"):
```
Comment (stephen) [inline internal/fields/sql.go]: nit: what is implicit about this null? i kind of like `zeroisnull`, but i could be convinced otherwise

File: sqlgen/implicitnull_test.go [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_byte": []byte("hello there")}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_time": timeField}, nil))
+
+			newRow.Id = 1
+			var rows []*ImplicitNull
+			require.NoError(t, db.Query(context.Background(), &rows, nil, nil))
+			assert.Equal(t, []*ImplicitNull{
+				newRow,
+			}, rows)
+		})
```
Comment (stephen) [inline sqlgen/implicitnull_test.go]: could we also add a test for the reading that a `null` in the db reads out as the zero value?

File: sqlgen/implicitnull_test.go (hunk 6 above)
Comment (willhug) [reply sqlgen/implicitnull_test.go]: 👍

File: internal/fields/reflect.go (hunk 8 above)
Comment (willhug) [reply internal/fields/reflect.go]: 👍

File: internal/fields/sql.go (hunk 10 above)
Comment (willhug) [reply internal/fields/sql.go]: No strong feelings either way, @berfarah @stephen whichever preference you guys want.

File: sqlgen/implicitnull_test.go (hunk 11 above)
Comment (willhug) [reply sqlgen/implicitnull_test.go]: 👍

File: internal/testfixtures/db.go (hunk 4 above)
Comment (willhug) [reply internal/testfixtures/db.go]: Yes

**Comment by willhug** [comment]
Not to self: binlog and Delete still need tests.
